### PR TITLE
Spectrum qnode transform

### PR DESCRIPTION
**Pull request by mariaschuld**
Adds the `spectrum` function.

### CODE REVIEW

File: pennylane/fourier/spectrum.py [hunk 1]
```diff
@@ -0,0 +1,244 @@
...
+    if g is None:
+        raise ValueError(f"no generator defined for operator {op}")
+
+    g = coeff * g
+    evals = np.linalg.eigvals(g)
+    # eigenvalues of hermitian ops are guaranteed to be real
+    evals = np.real(evals)
+
+    # append negative to cater for the complex conjugate part which subtracts eigenvalues
+    evals = [evals, -evals]
```
Comment (mariaschuld) [inline pennylane/fourier/spectrum.py:103]: I have to admit I am a bit confused why this is right here...I stole this from you

Comment (glassnotes) [reply pennylane/fourier/spectrum.py:103]: I believe this is from your original version of the function :sweat_smile:

File: pennylane/fourier/spectrum.py [hunk 2]
```diff
@@ -0,0 +1,244 @@
...
+        (Dict[pennylane.numpy.tensor, list[float]]): Dictionary of scalar inputs as keys, and
+        their frequency spectra as values.
+
+    **Example**
+
+    .. code-block:: python
+
+        import pennylane as qml
+        from pennylane import numpy as pnp
+        from pennylane.beta.transforms.fourier import spectrum
```
Comment (glassnotes) [inline pennylane/fourier/spectrum.py:176]:
```suggestion
        from pennylane.fourier import spectrum
```

File: pennylane/fourier/spectrum.py [hunk 3]
```diff
@@ -0,0 +1,244 @@
...
+    """
+
+    @wraps(qnode)
+    def wrapper(*args, **kwargs):
+
+        qnode_copy = copy(qnode)
+
+        # hack: currently the tape only differentiates trainable/non-trainable params
+        # if the qnode uses non-backprop diff rules.
+        qnode.diff_options["method"] = "parameter-shift"
```
Comment (glassnotes) [inline pennylane/fourier/spectrum.py:213]: Should this be `qnode_copy` here?

File: pennylane/fourier/spectrum.py [hunk 4]
```diff
@@ -0,0 +1,244 @@
...
+        >>> 0.2: [-2.0, -1.0, 0.0, 1.0, 2.0]
+        >>> 0.3: [-1.0, 0.0, 1.0]
+        >>> 0.4: [-2.0, -1.0, 0.0, 1.0, 2.0]
+        >>> 0.5: [-1.0, 0.0, 1.0]
+
+
+    """
+
+    @wraps(qnode)
+    def wrapper(*args, **kwargs):
```
Comment (glassnotes) [inline pennylane/fourier/spectrum.py:207]: Why is the wrapper necessary?

File: pennylane/fourier/spectrum.py [hunk 5]
```diff
@@ -0,0 +1,244 @@
...
+
+    # append negative to cater for the complex conjugate part which subtracts eigenvalues
+    evals = [evals, -evals]
+
+    frequencies = [np.round(sum(comb), decimals=8) for comb in product(*evals)]
+    unique_frequencies = list(set(frequencies))
+    return sorted(unique_frequencies)
+
+
+def _join_spectra(spec1, spec2):
```
Comment (glassnotes) [inline pennylane/fourier/spectrum.py:110]: Nice :+1:

File: tests/beta/test_transform_fourier.py [hunk 6]
```diff
@@ -0,0 +1,149 @@
...
+from pennylane import numpy as pnp
+from pennylane.fourier.spectrum import spectrum, _join_spectra, _get_spectrum, _simplify_tape
+
+
+class TestHelpers:
+
+    @pytest.mark.parametrize("spectrum1, spectrum2, expected", [([-1, 0, 1], [-1, 0, 1], [-2, -1, 0, 1, 2]),
+                                                                ([-3, 0, 3], [-5, 0, 5],
+                                                                 [-8, -5, -3, -2, 0, 2, 3, 5, 8]),
+                                                                ([-2, -1, 0, 1, 2], [-1, 0, 1],
```
Comment (glassnotes) [inline tests/beta/test_transform_fourier.py:29]: Unrelated (?) comment: these kinda remind me of the type of combinatorial designs called "difference sets".

File: pennylane/fourier/spectrum.py [hunk 7]
```diff
@@ -0,0 +1,244 @@
...
+
+    Args:
+        tape (~.tape.QuantumTape): tape to simplify
+        original_inputs (list): list of inputs, to check that they are conserved throughout expansion
+
+    Returns:
+        ~.tape.QuantumTape: expanded tape
+    """
+
+    def stop_at(obj):
```
Comment (glassnotes) [inline pennylane/fourier/spectrum.py:40]: Does this need to live inside the `_simplify_tape` function or can it be moved outside? It seems like it is most of the function.

Comment (mariaschuld) [reply pennylane/fourier/spectrum.py:40]: Oh yes, I forgot. Probably good to move it outside? I just thought the _simplify function is a good logical unit...

Let's worry about this in the final PR maybe!

File: pennylane/fourier/spectrum.py (hunk 1 above)
Comment (mariaschuld) [reply pennylane/fourier/spectrum.py:103]: Oh shoot :) No, I remember. Each gate appears in the left and right side of the quadratic form that makes up the expval!

File: pennylane/fourier/spectrum.py (hunk 4 above)
Comment (mariaschuld) [reply pennylane/fourier/spectrum.py:207]: I used the way other transforms do it...this allows you do have the syntax

```
qml.func(qnode)(qnode_args)
```

File: pennylane/fourier/spectrum.py (hunk 3 above)
Comment (mariaschuld) [reply pennylane/fourier/spectrum.py:213]: Good one!

Comment (mariaschuld) [inline pennylane/fourier/spectrum.py:213]:
```suggestion
        qnode_copy.diff_options["method"] = "parameter-shift"
```

**Pull request merged**